milatools/utils/vscode_utils.py

```diff
@@ -0,0 +1,487 @@
+from __future__ import annotations
+
+import functools
+import os
+import shlex
+import shutil
+import subprocess
+import sys
+import textwrap
+from logging import getLogger as get_logger
+from pathlib import Path
+from typing import Literal, Sequence
+
+from milatools.cli.local import Local
+from milatools.cli.remote import Remote
+from milatools.cli.utils import (
+    CLUSTERS,
+    batched,
+    stripped_lines_of,
+)
+from milatools.utils.parallel_progress import (
+    DictProxy,
+    ProgressDict,
+    TaskFn,
+    TaskID,
+    parallel_progress_bar,
+)
+from milatools.utils.remote_v2 import RemoteV2
+
+logger = get_logger(__name__)
+
+
+def running_inside_WSL() -> bool:
+    return sys.platform == "linux" and bool(shutil.which("powershell.exe"))
+
+
+def get_expected_vscode_settings_json_path() -> Path:
+    if sys.platform == "win32":
+        return Path.home() / "AppData\\Roaming\\Code\\User\\settings.json"
+    if sys.platform == "darwin":  # MacOS
+        return (
+            Path.home()
+            / "Library"
+            / "Application Support"
+            / "Code"
+            / "User"
+            / "settings.json"
+        )
+    if running_inside_WSL():
+        # Need to get the Windows Home directory, not the WSL one!
+        windows_username = subprocess.getoutput("powershell.exe '$env:UserName'")
+        return Path(
+            f"/mnt/c/Users/{windows_username}/AppData/Roaming/Code/User/settings.json"
+        )
+    # Linux:
+    return Path.home() / ".config/Code/User/settings.json"
+
+
+def get_code_command() -> str:
+    return os.environ.get("MILATOOLS_CODE_COMMAND", "code")
+
+
+def get_vscode_executable_path() -> str | None:
+    return shutil.which(get_code_command())
+
+
+def vscode_installed() -> bool:
+    return bool(get_vscode_executable_path())
+
+
+def sync_vscode_extensions_in_parallel_with_hostnames(
+    source: str,
+    destinations: list[str],
+):
+    destinations = list(destinations)
+    if source in destinations:
+        if source == "mila" and destinations == CLUSTERS:
+            logger.info("Assuming you want to sync from mila to all DRAC/CC clusters.")
+        else:
+            logger.warning(
+                f"{source=} is also in the destinations to sync to. " f"Removing it."
+            )
+        destinations.remove(source)
+
+    if len(set(destinations)) != len(destinations):
+        raise ValueError(f"{destinations=} contains duplicate hostnames!")
+
+    source_obj = Local() if source == "localhost" else RemoteV2(source)
+    return sync_vscode_extensions_in_parallel(source_obj, destinations)
+
+
+def sync_vscode_extensions_in_parallel(
```

Comment:
I think until we manage to run the ssh commands asynchronously (perhaps using something like [asyncssh](https://asyncssh.readthedocs.io/en/latest/)), then I am not aware of any other way that we could execute the commands over SSH to different hosts at the same time.

Reply:
Renamed the functions to `sync_vscode_extensions` in e88c025a607a053c4c1db572b0899f914f565518 just to avoid the implication that there is an equivalent synchronous function